# House Price Prediction Using Regularization for Surprise Housing

## Importing and Reading data

In [91]:
# Supress warnings
import warnings
warnings.filterwarnings('ignore')

In [92]:
# import python libraries
import pandas as pd
import numpy as np

In [93]:
# Importing dataset into 'train_df' python dataframe
train_df = pd.read_csv('Datasets/train.csv')
train_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [94]:
# Let us see the shape of the data
train_df.shape

(1460, 81)

In [95]:
# Let us check for the null values
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

We have 19 fields with Null values, let us print the field names.

In [96]:
train_df.columns[train_df.isna().any()].to_list()

['LotFrontage',
 'Alley',
 'MasVnrType',
 'MasVnrArea',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Electrical',
 'FireplaceQu',
 'GarageType',
 'GarageYrBlt',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PoolQC',
 'Fence',
 'MiscFeature']

Let us check the values in the field "Alley"

In [97]:
train_df.Alley.value_counts()

Grvl    50
Pave    41
Name: Alley, dtype: int64

Based on Data Description, Let us fill value 'No alley access' in the field "Alley".

In [98]:
train_df.Alley = train_df.Alley.fillna('No alley access')
train_df.Alley

0       No alley access
1       No alley access
2       No alley access
3       No alley access
4       No alley access
             ...       
1455    No alley access
1456    No alley access
1457    No alley access
1458    No alley access
1459    No alley access
Name: Alley, Length: 1460, dtype: object

0       No alley access
1       No alley access
2       No alley access
3       No alley access
4       No alley access
             ...       
1455    No alley access
1456    No alley access
1457    No alley access
1458    No alley access
1459    No alley access
Name: Alley, Length: 1460, dtype: object

In [99]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          1460 non-null   object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [100]:
# let us check for the values in the "MasVnrType" field
train_df.MasVnrType.value_counts()



# Pending......................

None       864
BrkFace    445
Stone      128
BrkCmn      15
Name: MasVnrType, dtype: int64

In [101]:
# let us check for the values in the "MasVnrArea" field
train_df.MasVnrArea.value_counts()



# Pending......................

0.0      861
180.0      8
72.0       8
108.0      8
120.0      7
        ... 
562.0      1
89.0       1
921.0      1
762.0      1
119.0      1
Name: MasVnrArea, Length: 327, dtype: int64

In [102]:
train_df.columns[train_df.isna().any()].to_list()

['LotFrontage',
 'MasVnrType',
 'MasVnrArea',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Electrical',
 'FireplaceQu',
 'GarageType',
 'GarageYrBlt',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PoolQC',
 'Fence',
 'MiscFeature']

In [103]:
# let us check for the values in the "BsmtQual" field
train_df.BsmtQual.value_counts()

TA    649
Gd    618
Ex    121
Fa     35
Name: BsmtQual, dtype: int64

In [104]:
# let us check for the values in the "BsmtCond" field
train_df.BsmtCond.value_counts()

TA    1311
Gd      65
Fa      45
Po       2
Name: BsmtCond, dtype: int64

Let us fill value 'No Basement' in the fields "BsmtQual" and "BsmtCond"

In [105]:
# Based on Data Description, Let us fill value 'NA' in the field "BsmtQual"
train_df.BsmtQual = train_df.BsmtQual.fillna('NA')
train_df.BsmtQual.value_counts()

TA    649
Gd    618
Ex    121
NA     37
Fa     35
Name: BsmtQual, dtype: int64

In [106]:
# Based on Data Description, Let us fill value 'NA' in the field "BsmtCond"
train_df.BsmtCond = train_df.BsmtCond.fillna('NA')
train_df.BsmtCond.value_counts()

TA    1311
Gd      65
Fa      45
NA      37
Po       2
Name: BsmtCond, dtype: int64

In [109]:
# let us check for the values in the "BsmtExposure" field
train_df.BsmtExposure.value_counts()

No    953
Av    221
Gd    134
Mn    114
Name: BsmtExposure, dtype: int64

In [119]:
# Based on Data Description, Let us fill value 'NA' in the field "BsmtExposure" for 'No Basement' 
train_df.BsmtExposure = train_df.BsmtExposure.fillna('NA')
train_df.BsmtExposure.value_counts()

No    953
Av    221
Gd    134
Mn    114
NA     38
Name: BsmtExposure, dtype: int64

In [122]:
# let us check for the values in the "BsmtFinType1" field
train_df.BsmtFinType1.value_counts()

Unf    430
GLQ    418
ALQ    220
BLQ    148
Rec    133
LwQ     74
Name: BsmtFinType1, dtype: int64

In [123]:
# Based on Data Description, Let us fill value 'NA' in the field "BsmtFinType1" for 'No Basement' 
train_df.BsmtFinType1 = train_df.BsmtFinType1.fillna('NA')
train_df.BsmtFinType1.value_counts()

Unf    430
GLQ    418
ALQ    220
BLQ    148
Rec    133
LwQ     74
NA      37
Name: BsmtFinType1, dtype: int64

In [124]:
# let us check for the values in the "BsmtFinType1" field
train_df.BsmtFinType2.value_counts()

Unf    1256
Rec      54
LwQ      46
BLQ      33
ALQ      19
GLQ      14
Name: BsmtFinType2, dtype: int64

In [125]:
# Based on Data Description, Let us fill value 'NA' in the field "BsmtFinType2" for 'No Basement' 
train_df.BsmtFinType2 = train_df.BsmtFinType2.fillna('NA')
train_df.BsmtFinType2.value_counts()

Unf    1256
Rec      54
LwQ      46
NA       38
BLQ      33
ALQ      19
GLQ      14
Name: BsmtFinType2, dtype: int64

In [132]:
# let us check for the values in the "Electrical" field
train_df.Electrical.value_counts()




# Pendng....

SBrkr    1334
FuseA      94
FuseF      27
FuseP       3
Mix         1
Name: Electrical, dtype: int64

In [133]:
# let us check for the values in the "FireplaceQu" field
train_df.FireplaceQu.value_counts()

Gd    380
TA    313
Fa     33
Ex     24
Po     20
Name: FireplaceQu, dtype: int64

In [134]:
# Based on Data Description, Let us fill value 'NA' in the field "BsmtFinType2" for 'No Fireplace' 
train_df.FireplaceQu = train_df.FireplaceQu.fillna('NA')
train_df.FireplaceQu.value_counts()

NA    690
Gd    380
TA    313
Fa     33
Ex     24
Po     20
Name: FireplaceQu, dtype: int64

In [137]:
# let us check for the values in the "GarageType" field
train_df.GarageType.value_counts()

Attchd     870
Detchd     387
BuiltIn     88
Basment     19
CarPort      9
2Types       6
Name: GarageType, dtype: int64

In [141]:
# Based on Data Description, Let us fill value 'NA' in the field "GarageType" for 'No Garage' 
train_df.GarageType = train_df.GarageType.fillna('NA')
train_df.GarageType.value_counts()

Attchd     870
Detchd     387
BuiltIn     88
NA          81
Basment     19
CarPort      9
2Types       6
Name: GarageType, dtype: int64

In [147]:
# let us check for the values in the "GarageYrBlt" field
min(train_df.GarageYrBlt)

1900.0

In [148]:
# let us check for the values in the "GarageYrBlt" field
max(train_df.GarageYrBlt)

2010.0

In [ ]:
# Based on Data Description, Let us fill value 'NA' in the field "GarageYrBlt" as there is no Garage.
train_df.GarageYrBlt = train_df.GarageYrBlt.fillna('NA')
train_df.GarageYrBlt.value_counts()




# Pending....

In [150]:
# let us check for the values in the "GarageFinish" field
train_df.GarageFinish.value_counts()

Unf    605
RFn    422
Fin    352
Name: GarageFinish, dtype: int64

In [151]:
# Based on Data Description, Let us fill value 'NA' in the field "GarageFinish" as there is no Garage.
train_df.GarageFinish = train_df.GarageFinish.fillna('NA')
train_df.GarageFinish.value_counts()

Unf    605
RFn    422
Fin    352
NA      81
Name: GarageFinish, dtype: int64

In [154]:
# let us check for the values in the "GarageQual" field
train_df.GarageQual.value_counts()

TA    1311
Fa      48
Gd      14
Ex       3
Po       3
Name: GarageQual, dtype: int64

In [156]:
# Based on Data Description, Let us fill value 'NA' in the field "GarageQual" as there is no Garage.
train_df.GarageQual = train_df.GarageQual.fillna('NA')
train_df.GarageQual.value_counts()

TA    1311
NA      81
Fa      48
Gd      14
Ex       3
Po       3
Name: GarageQual, dtype: int64

In [155]:
# let us check for the values in the "GarageCond" field
train_df.GarageCond.value_counts()

TA    1326
Fa      35
Gd       9
Po       7
Ex       2
Name: GarageCond, dtype: int64

In [157]:
# Based on Data Description, Let us fill value 'NA' in the field "GarageCond" as there is no Garage.
train_df.GarageCond = train_df.GarageCond.fillna('NA')
train_df.GarageCond.value_counts()

TA    1326
NA      81
Fa      35
Gd       9
Po       7
Ex       2
Name: GarageCond, dtype: int64

In [160]:
# let us check for the values in the "PoolQC" field
train_df.PoolQC.value_counts()

NA    1453
Gd       3
Ex       2
Fa       2
Name: PoolQC, dtype: int64

In [159]:
# Based on Data Description, Let us fill value 'NA' in the field "GarageCond" for 'No Pool'
train_df.PoolQC = train_df.PoolQC.fillna('NA')
train_df.PoolQC.value_counts()

NA    1453
Gd       3
Ex       2
Fa       2
Name: PoolQC, dtype: int64

In [162]:
# let us check for the values in the "Fence" field
train_df.Fence.value_counts()

MnPrv    157
GdPrv     59
GdWo      54
MnWw      11
Name: Fence, dtype: int64

In [163]:
# Based on Data Description, Let us fill value 'NA' in the field "GarageCond" for 'No Fence'
train_df.Fence = train_df.Fence.fillna('NA')
train_df.Fence.value_counts()

NA       1179
MnPrv     157
GdPrv      59
GdWo       54
MnWw       11
Name: Fence, dtype: int64

In [165]:
# let us check for the values in the "MiscFeature" field
train_df.MiscFeature.value_counts()

Shed    49
Gar2     2
Othr     2
TenC     1
Name: MiscFeature, dtype: int64

In [166]:
# Based on Data Description, Let us fill value 'NA' in the field "MiscFeature" for 'None'
train_df.MiscFeature = train_df.MiscFeature.fillna('NA')
train_df.MiscFeature.value_counts()

NA      1406
Shed      49
Gar2       2
Othr       2
TenC       1
Name: MiscFeature, dtype: int64

In [167]:
train_df.columns[train_df.isna().any()].to_list()

['LotFrontage', 'MasVnrType', 'MasVnrArea', 'Electrical', 'GarageYrBlt']

In [206]:
# Let us check for the minimum value in the field 'LotFrontage'
min(train_df.LotFrontage)

21.0

This means there is no 0 value in the field 'LotFrontage'. <br/>
Let us update the median value of field 'LotFrontage' for Null values.

In [209]:
LotFrontage_med = train_df.LotFrontage[train_df.LotFrontage!=0].median()
LotFrontage_med

69.0

In [219]:
# Let us check for the average of LotFrontage field.
train_df.LotFrontage[train_df.LotFrontage!=0].mean()

69.86369863013698

In [210]:
# Both median value and the average value of the field 'LotFrontage' is almost close, 
# hence, let us update above median value in the field 'LotFrontage'
train_df.LotFrontage = train_df.LotFrontage.fillna(LotFrontage_med)

In [212]:
# Let us check for the values in the "MasVnrArea" field and if it contains value 0
train_df.MasVnrArea.sort_values().value_counts()

0.0       861
180.0       8
108.0       8
72.0        8
16.0        7
         ... 
254.0       1
255.0       1
258.0       1
259.0       1
1600.0      1
Name: MasVnrArea, Length: 327, dtype: int64

Thus, there are 861 records with 0 vaulue. <br/>
Let us check for the non-zero values.

In [213]:
train_df.MasVnrArea[train_df.MasVnrArea!=0].sort_values().value_counts()

180.0     8
72.0      8
108.0     8
120.0     7
16.0      7
         ..
297.0     1
81.0      1
299.0     1
67.0      1
1600.0    1
Name: MasVnrArea, Length: 326, dtype: int64

In [217]:
# Let us check for the median of non-zero MasVnrArea values.
MasVnrArea_med = train_df.MasVnrArea[train_df.MasVnrArea!=0].median()
MasVnrArea_med

203.0

In [218]:
# Let us update above median value in the field 'MasVnrArea'
train_df.MasVnrArea = train_df.MasVnrArea.fillna(MasVnrArea_med)

In [168]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          1460 non-null   object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [131]:
train_df[train_df.Electrical.isna()==True]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street            Alley  \
1379  1380          80       RL         73.0     9735   Pave  No alley access   

     LotShape LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature  \
1379      Reg         Lvl    AllPub  ...        0    NaN   NaN         NaN   

     MiscVal MoSold YrSold  SaleType  SaleCondition  SalePrice  
1379       0      5   2008        WD         Normal     167500  

[1 rows x 81 columns]

In [140]:
train_df[train_df.GarageType.isna()==True]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street            Alley  \
39      40          90       RL         65.0     6040   Pave  No alley access   
48      49         190       RM         33.0     4456   Pave  No alley access   
78      79          90       RL         72.0    10778   Pave  No alley access   
88      89          50  C (all)        105.0     8470   Pave  No alley access   
89      90          20       RL         60.0     8070   Pave  No alley access   
...    ...         ...      ...          ...      ...    ...              ...   
1349  1350          70       RM         50.0     5250   Pave             Pave   
1407  1408          20       RL          NaN     8780   Pave  No alley access   
1449  1450         180       RM         21.0     1533   Pave  No alley access   
1450  1451          90       RL         60.0     9000   Pave  No alley access   
1453  1454          20       RL         90.0    17217   Pave  No alley access   

     LotShape LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature  \
39        Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
48        Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
78        Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
88        IR1         Lvl    AllPub  ...        0    NaN  MnPrv         NaN   
89        Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
...       ...         ...       ...  ...      ...    ...    ...         ...   
1349      Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
1407      IR1         Lvl    AllPub  ...        0    NaN  MnPrv         NaN   
1449      Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
1450      Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
1453      Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   

     MiscVal MoSold YrSold  SaleType  SaleCondition  SalePrice  
39         0      6   2008        WD        AdjLand      82000  
48         0      6   2009       New        Partial     113000  
78         0      4   2010        WD         Normal     136500  
88         0     10   2009     ConLD        Abnorml      85000  
89         0      8   2007        WD         Normal     123600  
...      ...    ...    ...       ...            ...        ...  
1349       0     12   2008        WD         Normal     122000  
1407       0      3   2009        WD         Normal     112000  
1449       0      8   2006        WD        Abnorml      92000  
1450       0      9   2009        WD         Normal     136000  
1453       0      7   2006        WD        Abnorml      84500  

[81 rows x 81 columns]

In [139]:
train_df[train_df.GarageYrBlt.isna()==True]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street            Alley  \
39      40          90       RL         65.0     6040   Pave  No alley access   
48      49         190       RM         33.0     4456   Pave  No alley access   
78      79          90       RL         72.0    10778   Pave  No alley access   
88      89          50  C (all)        105.0     8470   Pave  No alley access   
89      90          20       RL         60.0     8070   Pave  No alley access   
...    ...         ...      ...          ...      ...    ...              ...   
1349  1350          70       RM         50.0     5250   Pave             Pave   
1407  1408          20       RL          NaN     8780   Pave  No alley access   
1449  1450         180       RM         21.0     1533   Pave  No alley access   
1450  1451          90       RL         60.0     9000   Pave  No alley access   
1453  1454          20       RL         90.0    17217   Pave  No alley access   

     LotShape LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature  \
39        Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
48        Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
78        Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
88        IR1         Lvl    AllPub  ...        0    NaN  MnPrv         NaN   
89        Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
...       ...         ...       ...  ...      ...    ...    ...         ...   
1349      Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
1407      IR1         Lvl    AllPub  ...        0    NaN  MnPrv         NaN   
1449      Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
1450      Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   
1453      Reg         Lvl    AllPub  ...        0    NaN    NaN         NaN   

     MiscVal MoSold YrSold  SaleType  SaleCondition  SalePrice  
39         0      6   2008        WD        AdjLand      82000  
48         0      6   2009       New        Partial     113000  
78         0      4   2010        WD         Normal     136500  
88         0     10   2009     ConLD        Abnorml      85000  
89         0      8   2007        WD         Normal     123600  
...      ...    ...    ...       ...            ...        ...  
1349       0     12   2008        WD         Normal     122000  
1407       0      3   2009        WD         Normal     112000  
1449       0      8   2006        WD        Abnorml      92000  
1450       0      9   2009        WD         Normal     136000  
1453       0      7   2006        WD        Abnorml      84500  

[81 rows x 81 columns]